### ISSUE TITLE
Regression 4.0.2: Stack Overflow

### ISSUE THREAD

**Issue by peter-empen** [etorreborre/specs2#633]
Do you have a suggestion how to resolve the following issue that occurs on switching from from 3.8. to 4.0.2 and working extensively with the finder:

```
java.lang.StackOverflowError
	at org.specs2.fp.Traverse$$anon$1.$anonfun$traverseImpl$3(Traverse.scala:38)
	at org.specs2.fp.Monad.ap(Monad.scala:19)
	at org.specs2.fp.Monad.ap$(Monad.scala:18)
	at org.specs2.control.eff.EffImplicits$$anon$1.ap(Eff.scala:141)
	at org.specs2.fp.Applicative.ap2(Applicative.scala:18)
	at org.specs2.fp.Applicative.ap2$(Applicative.scala:17)
	at org.specs2.control.eff.EffImplicits$$anon$1.ap2(Eff.scala:141)
	at org.specs2.fp.Applicative.apply2(Applicative.scala:33)
	at org.specs2.fp.Applicative.apply2$(Applicative.scala:32)
	at org.specs2.control.eff.EffImplicits$$anon$1.apply2(Eff.scala:141)
	at org.specs2.fp.Traverse$$anon$1.traverseImpl(Traverse.scala:38)
	at org.specs2.fp.Traverse$$anon$1.$anonfun$traverseImpl$3(Traverse.scala:38)
```

**Comment by etorreborre** [comment]
Ouch, the first idea would be to run with `-Xss4m` in order to get a bigger stack.

**Comment by peter-empen** [comment]
Certainly. But, provided you can exclude an infinite loop, what about moving to a tail recursive implementation?

**Comment by etorreborre** [comment]
I suspect that the part where there's a loop is a part that I don't know how to make tail-recursive. I will try to at least put a work-around.

**Comment by peter-empen** [comment]
Yeh, that is often too hard a job. I sometimes also spent a lot of time to get specific algorithms tail-recursive.
Just for curiosity, could you point me to that method?

**Comment by etorreborre** [comment]
Sure, but just for curiosity then :-). Specs2 is running an old version of the eff library internally and its `ap` method for the `Eff` applicative is: https://github.com/etorreborre/specs2/blob/master/common/shared/src/main/scala/org/specs2/control/eff/Eff.scala#L177.

In the current version of `Eff` we have: https://github.com/atnos-org/eff/blob/master/shared/src/main/scala/org/atnos/eff/Eff.scala#L212 which is still not tail-recursive but has a built-in work-around for large traversals. I tried to implement another data structure with [this](https://github.com/atnos-org/eff/pull/118) but I am not convinced this is a good way forward. I should probably find some time to implement [something like this](https://github.com/typelevel/cats/pull/1748/files).

**Comment by etorreborre** [comment]
Quick question: if I want to reproduce your issue I should be able to do so by trying to find specifications in a very large directory (including nested directories)? Or did you use a directory containing many files?

**Comment by peter-empen** [comment]
Typically we have a depth of up to 10 with nodes having 1 to 6 descendants. Node directories tend to contain none or just a few files, while leaf directories contain up to 40 files.

**Comment by etorreborre** [comment]
ok thanks.

**Comment by etorreborre** [comment]
@peter-empen could you please try out `4.0.2-6a513a8-20180216164051` and tell me if that works for you? I have updated the Eff code and hopefully it will fix your problem.

**Comment by peter-empen** [comment]
Thank you, beware with me...

**Comment by peter-empen** [comment]
Finally here my test results: A StackOverflowError may still occur:
```
        at org.specs2.fp.Traverse$$anon$1.$anonfun$traverseImpl$3(Traverse.scala:43)
	at org.specs2.fp.Monad.ap(Monad.scala:19)
	at org.specs2.fp.Monad.ap$(Monad.scala:18)
	at org.specs2.control.eff.EffImplicits$$anon$1.ap(Eff.scala:151)
	at org.specs2.fp.Applicative.ap2(Applicative.scala:18)
	at org.specs2.fp.Applicative.ap2$(Applicative.scala:17)
	at org.specs2.control.eff.EffImplicits$$anon$1.ap2(Eff.scala:151)
	at org.specs2.fp.Applicative.apply2(Applicative.scala:33)
	at org.specs2.fp.Applicative.apply2$(Applicative.scala:32)
	at org.specs2.control.eff.EffImplicits$$anon$1.apply2(Eff.scala:151)
	at org.specs2.fp.Traverse$$anon$1.traverseImpl(Traverse.scala:43)
	at org.specs2.fp.Traverse$$anon$1.$anonfun$traverseImpl$3(Traverse.scala:43)
```

**Comment by etorreborre** [comment]
Hi @peter-empen sorry about that. I guess I need to find a good way to reproduce this issue before I can tackle it. Do you think you could create a repro project showing it? This will make me gain a huge amount of time which I don't really have at the moment :-).

**Comment by peter-empen** [comment]
Hi Eric, I could give it a try. But just to understanding the need:
What did you roughly do in "I have updated the Eff code"? You told you did not see how to make things tail recursive. If so, how do you plan to guarantee that it works with any directory structure?

**Comment by etorreborre** [comment]
My fix in "I have updated the Eff code" was about making traversals a bit more sequential (as in "Monadic") instead of potentially concurrent (with `Applicative` and `ap`). The trick is to implement `ap` by limiting the number of times it could recursively call itself. Sorry it's hard to explain and I think I need a concrete example to work on in order to fix it for good.

**Comment by peter-empen** [comment]
Did this change go to a regular release or is it solely in 4.0.2-6a513a8-20180216164051?

**Comment by etorreborre** [comment]
It hasn't gone yet to a regular release. I will probably do one soon, I'd like to fix a few more issues before that.

**Comment by djspiewak** [comment]
This *definitely* looks like the exact same issue we saw in #682, fwiw.

**Comment by etorreborre** [comment]
Yes I was meaning to ask. @peter-empen would you mind testing your project again with either `specs2-4.3.1` (for Scala 2.12, 2.11 is being published right now) or with `specs2-4.1.1`? Thanks.

**Comment by peter-empen** [comment]
After upgrading to 4.3.2, StackOverflowError has vanished fortunately. Thank you.
